### PR TITLE
Submitting Chatlog Project

### PR DESCRIPTION
**Pull request by tamikaxuross**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+      }
+      return message;
+    });
+    setChatMessages(updatedMessages);
+  };
+  const likedCount = chatMessages.filter((entry) => entry.liked).length;
   return (
     <div id="App">
       <header>
         <h1>Application title</h1>
```
Comment (anselrognlie) [inline src/App.jsx:22]: It would be nice to update this to at least a hard-coded representation of the conversation title: 
```
<h1>Chat between Vladimir and Estragon</h1>
```

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={entryClass}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:13]: Nice use of the supplied `TimeStamp`. We pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLike: PropTypes.func.isRequired,
+  isLocal: PropTypes.bool.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:29]: The `id` (omitted here), `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  return (
+    <div className="chat-log">
+      {entries.map((entry) => (
+        <ChatEntry
+          key={entry.id}
+          sender={entry.sender}
+          body={entry.body}
+          timeStamp={entry.timeStamp}
+          liked={entry.liked}
+          onLike={() => onToggleLike(entry.id)}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:17]: 🔎 In general, defining the callback wrapper in the parent component while iterating over the data solves the same problem as passing the callback and id into the child component and creating the event wrapper there. It even allows the `id` to be omitted (the child component calls the supplied function without needing to supply any parameters, since they are backed into the individualized callback it received).

However, we should note that in the tests for `ChatEntry`, the `id` _was_ supplied to the child component, implying that we are expecting it to be received and used by `ChatEntry`. Therefore, in this case, we should receive the message `id` as one of the props, and create the wrapper for the event handler here in the `ChatEntry`. This also mirrors the approaches we took for creating event handlers during class.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.string.isRequired,
+      liked: PropTypes.bool.isRequired,
+      onLike: PropTypes.func.isRequired,
+    })
+  ).isRequired,
+  onToggleLike: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:36]: Similar to the props for `ChatEntry`, here, the `entries` prop is included in the tests, but the like toggle is not (both per message and at the log level), resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
+
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+
+const ChatLog = ({ entries, onToggleLike }) => {
+  const localSender = 'Vladimir';
+  return (
+    <div className="chat-log">
+      {entries.map((entry) => (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:10]: Nice use of `map` to convert from the message data into `ChatEntry` components. We could perform this mapping storing the result into a variable we use in the JSX result, we could make a helper function that we call as part of the `return`, or we can perform the map expression as part of the `return` JSX as you did here. I often like having relatively simple maps directly in the JSX since it helps me see the overall structure of the component, though it can make debugging a little more tricky. But any of those approaches will work fine.

File: src/components/ChatEntry.jsx [hunk 7]
```diff
@@ -1,20 +1,32 @@
...
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
+        </p>
+        <button className="like" onClick={onLike}>
+          {liked ? '❤️' : '🤍'}
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:16]: 👍 Nice use of a ternary to pick the emoji to show for the liked status. Consider moving even something small like this to a helper function external to the component, so that the behavior of the emoji-picking logic can be tested easily outside of React.

File: src/App.jsx [hunk 8]
```diff
@@ -1,14 +1,29 @@
...
+  const [chatMessages, setChatMessages] = useState(messages);
+
+  const toggleLike = (id) => {
+    const updatedMessages = chatMessages.map((message) => {
+      if (message.id === id) {
+        return { ...message, liked: !message.liked };
+      }
+      return message;
+    });
+    setChatMessages(updatedMessages);
```
Comment (anselrognlie) [inline src/App.jsx:16]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setChatMessages(chatMessages => chatMessages.map(message => {
      if (message.id === id) {
        return {...message, liked: !message.liked};
      } else {
        return message;
      };
    }));
```
We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,29 @@
...
+  };
+  const likedCount = chatMessages.filter((entry) => entry.liked).length;
```
Comment (anselrognlie) [inline src/App.jsx:18]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Great idea to filter down the data to the liked messages and use the list length as the count. This is a really understandable alternative to the more canonical `reduce` approach. Compared to `reduce`, it does make a list potentially as long as the message list, but since we know a similar copying gets done anyway during a render (building the `ChatEntry` components) on the whole it's not any more costly.

File: src/components/ChatEntry.jsx [hunk 10]
```diff
@@ -1,20 +1,32 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
 
-const ChatEntry = () => {
+const ChatEntry = ({ sender, body, timeStamp, liked, onLike, isLocal }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params), but I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 11]
```diff
@@ -1,20 +1,32 @@
...
+  const entryClass = isLocal ? 'chat-entry local' : 'chat-entry remote';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: 👍 Abstracting whether the entry is local or remote means the component can just be told whether it is local without requiring any logic (beyond translating from the local value to the class to apply).

File: src/components/ChatLog.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
+
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+
+const ChatLog = ({ entries, onToggleLike }) => {
+  const localSender = 'Vladimir';
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:7]: How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatLog` itself have enough information as it is to "know" which messages are local and which are remote? 

Perhaps the App could calculate which sender(s) are local and remote, and include that information as an additional prop?

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
